### PR TITLE
[SPARK-41914][SQL] FileFormatWriter materializes AQE plan before accessing outputOrdering

### PR DESCRIPTION
**Pull request by EnricoMi**
### What changes were proposed in this pull request?
The `FileFormatWriter` materializes an `AdaptiveQueryPlan` before accessing the plan's `outputOrdering`. This is required when planned writing is disabled (`spark.sql.optimizer.plannedWrite.enabled` is `true` by default). With planned writing enabled `FileFormatWriter` gets the final plan already.

### Why are the changes needed?
`FileFormatWriter` enforces an ordering if the written plan does not provide that ordering. An `AdaptiveQueryPlan` does not know its final ordering, in which case `FileFormatWriter` enforces the ordering (e.g. by column `"a"`) even if the plan provides a compatible ordering (e.g. by columns `"a", "b"`). In case of spilling, that order (e.g. by columns `"a", "b"`) gets broken (see SPARK-40588).

### Does this PR introduce _any_ user-facing change?
This fixes SPARK-40588 for 3.4, which was introduced in 3.0. This restores behaviour from Spark 2.4.

### How was this patch tested?
The final plan that is written to files is now stored in `FileFormatWriter.executedPlan` (similar to existing `FileFormatWriter.outputOrderingMatched`). Unit tests assert the outermost sort order written to files.

The actual plan written into the files changed from (taken from `"SPARK-41914: v1 write with AQE and in-partition sorted - non-string partition column"`):

```
Sort [input[2, int, false] ASC NULLS FIRST], false, 0
+- *(3) Sort [key#13 ASC NULLS FIRST, value#14 ASC NULLS FIRST], false, 0
   +- *(3) Project [b#24, value#14, key#13]
      +- *(3) BroadcastHashJoin [key#13], [a#23], Inner, BuildLeft, false
         :- BroadcastQueryStage 2
         :  +- BroadcastExchange HashedRelationBroadcastMode(List(cast(input[0, int, false] as bigint)),false), [plan_id=376]
         :     +- AQEShuffleRead local
         :        +- ShuffleQueryStage 0
         :           +- Exchange hashpartitioning(key#13, 5), ENSURE_REQUIREMENTS, [plan_id=328]
         :              +- *(1) SerializeFromObject [knownnotnull(assertnotnull(input[0, org.apache.spark.sql.test.SQLTestData$TestData, true])).key AS key#13, staticinvoke(class org.apache.spark.unsafe.types.UTF8String, StringType, fromString, knownnotnull(assertnotnull(input[0, org.apache.spark.sql.test.SQLTestData$TestData, true])).value, true, false, true) AS value#14]
         :                 +- Scan[obj#12]
         +- AQEShuffleRead local
            +- ShuffleQueryStage 1
               +- Exchange hashpartitioning(a#23, 5), ENSURE_REQUIREMENTS, [plan_id=345]
                  +- *(2) SerializeFromObject [knownnotnull(assertnotnull(input[0, org.apache.spark.sql.test.SQLTestData$TestData2, true])).a AS a#23, knownnotnull(assertnotnull(input[0, org.apache.spark.sql.test.SQLTestData$TestData2, true])).b AS b#24]
                     +- Scan[obj#22]
```

where `FileFormatWriter` enforces order with `Sort [input[2, int, false] ASC NULLS FIRST], false, 0`, to

```
*(3) Sort [key#13 ASC NULLS FIRST, value#14 ASC NULLS FIRST], false, 0
+- *(3) Project [b#24, value#14, key#13]
   +- *(3) BroadcastHashJoin [key#13], [a#23], Inner, BuildLeft, false
      :- BroadcastQueryStage 2
      :  +- BroadcastExchange HashedRelationBroadcastMode(List(cast(input[0, int, false] as bigint)),false), [plan_id=375]
      :     +- AQEShuffleRead local
      :        +- ShuffleQueryStage 0
      :           +- Exchange hashpartitioning(key#13, 5), ENSURE_REQUIREMENTS, [plan_id=327]
      :              +- *(1) SerializeFromObject [knownnotnull(assertnotnull(input[0, org.apache.spark.sql.test.SQLTestData$TestData, true])).key AS key#13, staticinvoke(class org.apache.spark.unsafe.types.UTF8String, StringType, fromString, knownnotnull(assertnotnull(input[0, org.apache.spark.sql.test.SQLTestData$TestData, true])).value, true, false, true) AS value#14]
      :                 +- Scan[obj#12]
      +- AQEShuffleRead local
         +- ShuffleQueryStage 1
            +- Exchange hashpartitioning(a#23, 5), ENSURE_REQUIREMENTS, [plan_id=344]
               +- *(2) SerializeFromObject [knownnotnull(assertnotnull(input[0, org.apache.spark.sql.test.SQLTestData$TestData2, true])).a AS a#23, knownnotnull(assertnotnull(input[0, org.apache.spark.sql.test.SQLTestData$TestData2, true])).b AS b#24]
                  +- Scan[obj#22]
```

where the sort given by the user is the outermost sort now.

### CODE REVIEW

**Comment by EnricoMi** [comment]
@cloud-fan here is fix for SPARK-40588 migrated to Spark 3.4.

This finally includes unit tests for the actual plan written to files (that has never been tested before).

File: sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/V1WriteCommandSuite.scala [hunk 1]
```diff
@@ -181,13 +178,111 @@ class V1WriteCommandSuite extends QueryTest with SharedSparkSession with V1Write
             |PARTITIONED BY (k STRING)
             |""".stripMargin)
         executeAndCheckOrdering(
-          hasLogicalSort = true, orderingMatched = enabled, hasEmpty2Null = enabled) {
+          hasLogicalSort = true, orderingMatched = true, hasEmpty2Null = enabled) {
```
Comment (cloud-fan) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/V1WriteCommandSuite.scala:181]: do we still need the `orderingMatched` parameter if it's always true?

File: sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/V1WriteCommandSuite.scala [hunk 2]
```diff
@@ -181,13 +178,111 @@ class V1WriteCommandSuite extends QueryTest with SharedSparkSession with V1Write
...
+          case s: SortExec => s
+        }.map(s => (enabled, s)).exists {
+          case (false, SortExec(Seq(
+          SortOrder(AttributeReference("value", StringType, _, _), Ascending, NullsFirst, _),
+          SortOrder(AttributeReference("key", IntegerType, _, _), Ascending, NullsFirst, _)
+          ), false, _, _)) => true
+
+          // SPARK-40885: this bug removes the in-partition sort, which manifests here
+          case (true, SortExec(Seq(
+          SortOrder(AttributeReference("value", StringType, _, _), Ascending, NullsFirst, _)
```
Comment (cloud-fan) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/V1WriteCommandSuite.scala:278]: do you know why the sorting key is different when planned write is enabled?

Comment (EnricoMi) [reply sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/V1WriteCommandSuite.scala:278]: that is correctness bug SPARK-40885 discussed in #38356

File: sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/V1WriteCommandSuite.scala (hunk 1 above)
Comment (EnricoMi) [reply sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/V1WriteCommandSuite.scala:181]: Not sure what you mean, as in needed in this test?

Comment (cloud-fan) [reply sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/V1WriteCommandSuite.scala:181]: I mean, shall we remove `orderingMatched` from the method `executeAndCheckOrdering`?

Comment (EnricoMi) [reply sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/V1WriteCommandSuite.scala:181]: There is no default value for `orderingMatched` and two other unit tests still use `orderingMatched=enabled`.

**Comment by EnricoMi** [comment]
All tests green: https://github.com/G-Research/spark/actions/runs/3855300306

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/FileFormatWriter.scala [hunk 3]
```diff
@@ -138,9 +144,21 @@ object FileFormatWriter extends Logging {
...
+    def materializeAdaptiveSparkPlan(plan: SparkPlan): SparkPlan = plan match {
+      case a: AdaptiveSparkPlanExec => a.finalPhysicalPlan
+      case p: SparkPlan => p.withNewChildren(p.children.map(materializeAdaptiveSparkPlan))
+    }
+
     // the sort order doesn't matter
     // Use the output ordering from the original plan before adding the empty2null projection.
-    val actualOrdering = writeFilesOpt.map(_.child).getOrElse(plan).outputOrdering.map(_.child)
+    val actualOrdering = writeFilesOpt.map(_.child)
+      .getOrElse(materializeAdaptiveSparkPlan(plan))
```
Comment (ulysses-you) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/FileFormatWriter.scala:160]: shall we put all code changes inside if `writeFilesOpt` is empty ? if `writeFilesOpt` is defined that means the write have been planned which does not have this issue.

Comment (cloud-fan) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/FileFormatWriter.scala:160]: `.getOrElse` already does what you said, isn't it?

File: sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/V1WriteCommandSuite.scala (hunk 2 above)
Comment (cloud-fan) [reply sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/V1WriteCommandSuite.scala:278]: @ulysses-you can you take a look at this bug?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/FileFormatWriter.scala (hunk 3 above)
Comment (EnricoMi) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/FileFormatWriter.scala:160]: Yes, `materializeAdaptiveSparkPlan` is applied on `plan` _only_ if `writeFilesOpt` is undefined.

**Comment by cloud-fan** [comment]
thanks, merging to master!

**Comment by EnricoMi** [comment]
Thanks!

**Pull request closed without merging**